lib/fluent/supervisor.rb

```diff
@@ -346,64 +372,73 @@ def stop_rpc_server
       @rpc_server.shutdown
     end
 
-    def supervise(&block)
-      start_time = Time.now
-
-      Process.setproctitle("supervisor:#{@process_name}") if @process_name
+    def supervise
       $log.info "starting fluentd-#{Fluent::VERSION}"
 
-      if !Fluent.windows?
-        @main_pid = fork do
-          main_process(&block)
-        end
+      if ServerEngine.windows?
+        fluentd_spawn_cmd = ServerEngine.ruby_bin_path + ' ' + $0 + ' '
       else
-        if @supervise
-          fluentd_spawn_cmd = @rubybin_dir+"/ruby.exe '"+@rubybin_dir+"/fluentd' "
-          $fluentdargv.each{|a|
-            fluentd_spawn_cmd << (a + " ")
-          }
-          fluentd_spawn_cmd << ("--no-supervisor")
-          $log.info "spawn command to main (windows) : " + fluentd_spawn_cmd
-          @main_pid = Process.spawn(fluentd_spawn_cmd)
-        else
-          main_process(&block)
-        end
-      end
-
-      if @daemonize && @wait_daemonize_pipe_w
-        STDIN.reopen("/dev/null")
-        STDOUT.reopen("/dev/null", "w")
-        STDERR.reopen("/dev/null", "w")
-        @wait_daemonize_pipe_w.close
-        @wait_daemonize_pipe_w = nil
+        fluentd_spawn_cmd = $0 + ' '
       end
 
-      Process.waitpid(@main_pid)
-      @main_pid = nil
-      ecode = $?.to_i
+      $fluentdargv.each{|a|
+        fluentd_spawn_cmd << (a + " ")
+      }
+      fluentd_spawn_cmd << (" --no-supervisor")
+      $log.info "spawn command to main: " + fluentd_spawn_cmd
+
+      params = {}
+      params['main_cmd'] = fluentd_spawn_cmd
+      params['daemonize'] = @daemonize
+      params['inline_config'] = @inline_config
+      params['log_path'] = @log_path
+      params['log_level'] = @log_level
+      params['chuser'] = @chuser
+      params['chgroup'] = @chgroup
+      params['use_v1_config'] = @use_v1_config
+      params['suppress_repeated_stacktrace'] = @suppress_repeated_stacktrace
+
+      se = ServerEngine.create(ServerModule, WorkerModule){
+        Fluent.load_config(@config_path, params)
+      }
+      se.run
+    end
 
-      if Fluent.windows?
-        @th_sv.kill if @th_sv.alive?
-        @th_sv.join rescue nil
+    def install_main_process_signal_handlers
+      # When user use Ctrl + C not SIGINT, SIGINT is sent to all process in same process group.
+      # Then serverengine can't handle signal, so need to handle it in this process.
+      trap :INT do
+        $log.debug "fluentd main process get SIGINT"
+        unless @finished
+          @finished = true
+          $log.debug "getting start to shutdown main process"
+          Fluent::Engine.stop
+        end
       end
 
-      $log.info "process finished", code: ecode
+      trap :USR1 do
```

Comment:
We want to support force flushing of buffers even in Windows by using RPC server calls, but with this implementation, it can't be done.
This actual procedure should be written in a method, and signal handler (and supervisor-worker RPC handler or so for windows) should call it.